pkg/auth/rbac.go

```diff
@@ -0,0 +1,353 @@
+)
+
+type Rbac struct {
+	policyStore *PolicyStore
+	workload    *workloadapi.Workload
+}
+
+func NewRbac() *Rbac {
+	return &Rbac{
+		policyStore: newPolicystore(),
+		workload:    &workloadapi.Workload{},
+	}
+}
+
+func (rbac *Rbac) DoRbac(conn *RbacConnection) bool {
+	allowPolices, denyPolicies := rbac.aggregate()
+
+	// 1. If there is ANY deny policy, deny the request
+	for _, denyPolicy := range denyPolicies {
+		if matches(conn, denyPolicy) {
+			return false
+		}
+	}
+
+	// 2. If there is NO allow policy for the workload, allow the request
+	if len(allowPolices) == 0 {
+		return true
+	}
+
+	// 3. If there is ANY allow policy matched, allow the request
+	for _, allowPolicy := range allowPolices {
+		if matches(conn, allowPolicy) {
+			return true
+		}
+	}
+
+	// 4. If 1,2 and 3 unsatisfied, deny the request
+	return false
+}
+
+func (rbac *Rbac) UpdatePolicy(auth *security.Authorization) error {
+	return rbac.policyStore.updatePolicy(auth)
+}
+
+func (rbac *Rbac) RemovePolicy(policyKey string) error {
+	return rbac.policyStore.removePolicy(policyKey)
+}
+
+func (rbac *Rbac) aggregate() (allowPolicies, denyPolicies []*AuthPolicy) {
+	// Collect policy names from workload, global namespace and namespace
+	policyNames := make([]string, len(rbac.workload.GetAuthorizationPolicies()))
+	copy(policyNames, rbac.workload.GetAuthorizationPolicies())
+	policyNames = append(policyNames, strSetToSlice(rbac.policyStore.byNamespace[""])...)
+	policyNames = append(policyNames, strSetToSlice(rbac.policyStore.byNamespace[rbac.workload.Namespace])...)
+
+	allowPolicies = make([]*AuthPolicy, 0)
+	denyPolicies = make([]*AuthPolicy, 0)
+	for _, name := range policyNames {
+		if policy, ok := rbac.policyStore.byKey[name]; !ok {
+			continue
+		} else {
+			if policy.Action == security.Action_ALLOW {
+				allowPolicies = append(allowPolicies, policy)
+			} else if policy.Action == security.Action_DENY {
+				denyPolicies = append(denyPolicies, policy)
+			}
+		}
+	}
+	return
+}
+
+func matches(conn *RbacConnection, policy *AuthPolicy) bool {
+	if policy.GetRules() == nil {
+		return false
+	}
+
+	eg := &errgroup.Group{}
+	mut := &sync.Mutex{}
+	// If ANY rule matches, it's a match
+	for _, rule := range policy.GetRules() {
+		ruleMatch := true
+		// If ALL clause matches, it's a match
+		for _, clause := range rule.GetClauses() {
+			clauseMatch := false
+			// If ANY match matches, it's a match
+			for _, match := range clause.GetMatches() {
+				if isEmptyMatch(match) {
+					continue
+				}
+				anyMatch := false
+
+				eg.Go(func() error {
+					m := matchDstIp(conn.DstIp, match)
+					mut.Lock()
+					anyMatch = anyMatch || m
+					mut.Unlock()
+					return nil
+				})
+
+				eg.Go(func() error {
+					m := matchSrcIp(conn.SrcIp, match)
+					mut.Lock()
+					anyMatch = anyMatch || m
+					mut.Unlock()
+					return nil
+				})
+
+				eg.Go(func() error {
+					m := matchDstPort(conn.DstPort, match)
+					mut.Lock()
+					anyMatch = anyMatch || m
```

Comment:
Should  exit early when meet a match?

Reply:
If multiple types are set, they are AND-ed, so the logic here turns out to AND, then no need to have an early exit